cargo-ndk/src/apk.rs

```diff
@@ -0,0 +1,113 @@
+use crate::manifest::Manifest;
+use android_build_tools::apk::{Apk, ApkConfig};
+use android_build_tools::config::Config;
+use android_build_tools::error::NdkError;
+use android_build_tools::ndk::Ndk;
+use android_build_tools::target::Target;
+use cargo_subcommand::{Artifact, CrateType, Subcommand};
+use std::path::PathBuf;
+use std::process::Command;
+
+pub struct ApkBuilder<'a> {
+    cmd: &'a Subcommand,
+    ndk: Ndk,
+    manifest: Manifest,
+    build_dir: PathBuf,
+    build_targets: Vec<Target>,
+}
+
+impl<'a> ApkBuilder<'a> {
+    pub fn from_subcommand(cmd: &'a Subcommand) -> Result<Self, NdkError> {
+        let ndk = Ndk::from_env()?;
+        let manifest = Manifest::parse_from_toml(cmd.manifest())?;
+        let build_targets = if let Some(target) = cmd.target() {
+            vec![Target::from_rust_triple(target)?]
+        } else {
+            vec![ndk.detect_abi().unwrap_or(Target::Arm64V8a)]
+        };
+        let build_dir = cmd.target_dir().join(cmd.profile()).join("ndk");
+        Ok(Self {
+            cmd,
+            ndk,
+            manifest,
+            build_dir,
+            build_targets,
+        })
+    }
+
+    pub fn cargo_build(&self, ndk: &Ndk, target: Target, sdk_version: u32) -> Result<(), NdkError> {
+        let triple = target.rust_triple();
+        let mut cargo = Command::new("cargo");
+        cargo.arg("build");
+
+        let clang = ndk.clang(target, sdk_version)?;
+        cargo.env(format!("CC_{}", triple), &clang);
+        cargo.env(cargo_env_target_cfg("LINKER", triple), &clang);
+
+        let ar = ndk.toolchain_bin("ar", target)?;
+        cargo.env(format!("AR_{}", triple), &ar);
+        cargo.env(cargo_env_target_cfg("AR", triple), &ar);
+
+        if self.cmd.target().is_none() {
+            cargo.arg("--target").arg(triple);
+        }
+
+        cargo.args(self.cmd.args()).status()?;
+        Ok(())
+    }
+
+    pub fn build(&self, artifact: &Artifact) -> Result<Apk, NdkError> {
+        let config = Config {
+            ndk: self.ndk.clone(),
+            build_dir: self.build_dir.join(artifact),
+            package_name: package_name(artifact),
+            package_label: artifact.name().to_string(),
+            version_name: self.manifest.version.clone(),
+            version_code: 1,
```

Comment:
According to the [docs](https://developer.android.com/guide/topics/manifest/manifest-element.html#vcode) this number is used to to determine that one version is more recent than another. So I think it should be either set directly from metadata or generated in more complicated way.

Reply:
Ok.

Reply:
How about using `<version_code> = <apk_id><major><minor><patch>` as the default? That would work for most projects, where the project can compute it's `<apk_id>` as the product of it's build matrix: `arch * dpi * target_sdk_version`

Reply:
LGTM.